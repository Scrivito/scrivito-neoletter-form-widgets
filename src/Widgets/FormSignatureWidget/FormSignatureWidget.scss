.scrivito-neoletter-form-widgets .form-signature {
  p {
    display: inline;
  }
  .signature {
    button.delete {
      cursor: pointer;
    }
    canvas.neoletter-form-canvas {
      background: white;
      border: 2px solid black;
      width: 100%;
      aspect-ratio: 5 / 2;
    }
  }
}
