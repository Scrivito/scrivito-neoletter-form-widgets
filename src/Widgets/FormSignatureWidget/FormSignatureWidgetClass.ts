import * as Scrivito from "scrivito";

export const FormSignatureWidget = Scrivito.provideWidgetClass("FormSignatureWidget", {
  attributes: {
    title: "html",
    helpText: "html",
    customFieldName: "string",
    strokeThickness: "integer",
    strokeColor: "string",
    backgroundColor: "string",
    deleteButtonText: "string",
    deleteButtonAlignment: [
      "enum",
      {
        values: ["left", "text-center", "text-end", "block"]
      }
    ]
  }
});
