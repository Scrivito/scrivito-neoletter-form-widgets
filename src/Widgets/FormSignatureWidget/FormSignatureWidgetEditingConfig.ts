import * as Scrivito from "scrivito";
import { customFieldNameValidation } from "../FormStepContainerWidget/utils/validations/customFieldNameValidation";
import { insideFormContainerValidation } from "../FormStepContainerWidget/utils/validations/insideFormContainerValidation";
import formSignatureWidgetIcon from "../../assets/images/form-signature-widget.svg";

Scrivito.provideEditingConfig("FormSignatureWidget", {
  title: "Neoletter Form Signature",
  thumbnail: formSignatureWidgetIcon,
  attributes: {
    title: { title: "Label" },
    customFieldName: { title: "Field name" },
    strokeThickness: { title: "Stroke thickness", description: 'Default: "2"' },
    strokeColor: { title: "Stroke color", description: 'Default: "black"' },
    backgroundColor: { title: "Background color", description: 'Default: "#ffffff"' },
    deleteButtonText: { title: "Delete button text" },
    deleteButtonAlignment: {
      title: "Delete button alignment",
      values: [
        { value: "left", title: "Left" },
        { value: "text-center", title: "Center" },
        { value: "text-end", title: "Right" },
        { value: "block", title: "Full width" }
      ]
    },
    helpText: { title: "Help text" }
  },
  properties: ["title", "customFieldName", "strokeThickness", "strokeColor", "backgroundColor", "deleteButtonText", "deleteButtonAlignment", "helpText"],
  initialContent: {
    title: "Sign here",
    customFieldName: "custom_",
    deleteButtonText: "Delete",
    strokeThickness: 2,
    strokeColor: "black",
    backgroundColor: "#ffffff",
    deleteButtonAlignment: "left"

  },
  validations: [insideFormContainerValidation, customFieldNameValidation]
});
