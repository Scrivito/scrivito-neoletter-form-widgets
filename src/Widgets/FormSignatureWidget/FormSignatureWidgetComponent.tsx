import * as React from "react";
import * as Scrivito from "scrivito";
import { HelpText } from "../FormStepContainerWidget/components/HelpTextComponent";
import { FormSignatureWidget } from "./FormSignatureWidgetClass";
import { getFieldName } from "../FormStepContainerWidget/utils/getFieldName";
import { Signature } from "../FormStepContainerWidget/components/SignatureComponent";
import "./FormSignatureWidget.scss";
Scrivito.provideComponent(FormSignatureWidget, ({ widget }) => {
  const [signatureDataUrl, setSignatureDataUrl] = React.useState<string>("")

  const onChangeSignature = (dataUrl: string) => {
    setSignatureDataUrl(dataUrl);
  }

  return (
    <div className="form-signature mb-3">
      {widget.get("title") && <div className="signature-title">
        <Scrivito.ContentTag
          attribute="title"
          content={widget}
          tag="span"
        />
        {widget.get("helpText") && <HelpText widget={widget} />}
      </div>
      }
      <Signature
        id={widget.id()}
        onChange={onChangeSignature}
        deleteButtonText={widget.get("deleteButtonText") || "Delete"}
        strokeColor={widget.get("strokeColor") || "black"}
        strokeThickness={widget.get("strokeThickness") || 2}
        backgroundColor={widget.get("backgroundColor") || "#ffffff"}
        deleteButtonAlignment={widget.get("deleteButtonAlignment") || "block"}
      ></Signature>
      <input
        type="hidden"
        name={getFieldName(widget)}
        value={signatureDataUrl}></input>
    </div>
  );
});
