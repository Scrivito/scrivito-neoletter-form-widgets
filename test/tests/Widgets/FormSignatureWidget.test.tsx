import * as Scrivito from "scrivito";
import 'jest-canvas-mock';
import { fireEvent } from "@testing-library/react";
import { FormSignatureWidget } from "../../../src/Widgets/FormSignatureWidget/FormSignatureWidgetClass";
import "../../../src/Widgets/FormSignatureWidget/FormSignatureWidgetComponent";
import PageRenderer from "../../helpers/pageRenderer";

Scrivito.configure({ tenant: "inMemory" });

const pageRenderer = new PageRenderer();

describe("FormSignatureWidget", () => {
	const widgetProps = {
		title: "Signature Title",
		helpText: "<p>This is a help text</p>",
		deleteButtonText: "Clear Signature",
		strokeColor: "black",
		strokeThickness: 2,
		backgroundColor: "#ffffff",
		deleteButtonAlignment: "block"
	};

	it("handles signature change and updates hidden input value", () => {
		pageRenderer.render({
			body: [new FormSignatureWidget(widgetProps)]
		});

		const signatureCanvas = document.querySelector("canvas")!;
		const hiddenInput: HTMLInputElement = document.querySelector('input[type="hidden"]')!;

		expect(signatureCanvas).toBeInTheDocument();
		expect(hiddenInput).toBeInTheDocument();
		expect(hiddenInput.value).toBe("");

		const emptyBase64String = "data:image/png;base64,00";
		expect(signatureCanvas.toDataURL()).toMatch(emptyBase64String);


		fireEvent.mouseDown(signatureCanvas, { clientX: 50, clientY: 50 });
		fireEvent.mouseMove(signatureCanvas, { clientX: 100, clientY: 100 });
		fireEvent.mouseUp(signatureCanvas);

		const ctx = signatureCanvas.getContext("2d")!;
		const base64Result = signatureCanvas.toDataURL();

		expect(ctx.moveTo).toHaveBeenCalled();
		expect(ctx.lineTo).toHaveBeenCalled();
		expect(ctx.stroke).toHaveBeenCalled();

		expect(hiddenInput.value).toBe(base64Result);
	});

	it("renders correctly", () => {
		const tree = pageRenderer.getAsJSON({
			body: [new FormSignatureWidget(widgetProps)]
		});
		expect(tree).toMatchSnapshot();
	});
});
