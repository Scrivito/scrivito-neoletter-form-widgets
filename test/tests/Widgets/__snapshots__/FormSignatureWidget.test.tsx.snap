// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSignatureWidget renders correctly 1`] = `
<div>
  <div>
    <div
      className="form-signature mb-3"
    >
      <div
        className="signature-title"
      >
        <span
          dangerouslySetInnerHTML={
            {
              "__html": "Signature Title",
            }
          }
          onClick={[Function]}
        />
        <div
          className="helptext-container"
        >
          <i
            className="bi bi-question-circle"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          />
        </div>
      </div>
      <div
        className="signature"
        id="staticWidgetID"
      >
        <canvas
          className="neoletter-form-canvas"
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseOut={[Function]}
          onMouseUp={[Function]}
          onTouchCancel={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
        />
        <div
          className=""
        >
          <button
            className="delete btn btn-primary btn-block"
            onClick={[Function]}
          >
            <span
              className="delete-text"
            >
              Clear Signature
            </span>
          </button>
        </div>
      </div>
      <input
        name=""
        type="hidden"
        value=""
      />
    </div>
  </div>
</div>
`;
